### PR TITLE
[PIP-158][improve][client] Split client TLS transport encryption from authentication

### LINKED ISSUE #15289: PIP-158: Split client TLS transport encryption from authentication
Discussion thread: https://lists.apache.org/thread/fblmm8oc7h907cfnppvk71o2cbp5mk8q
## Motivation

The client supports TLS transport encryption and TLS authentication, this code so like:

```java
PulsarClient client = PulsarClient.builder()
                .serviceUrl("pulsar+ssl://localhost:6651")
                .tlsTrustCertsFilePath("/path/to/cacert.pem")
                .authentication(AuthenticationTls.class.getName(), authParams)
                .build()
```

We are using AuthenticationTls to perform the TLS transport encryption and TLS authentication, this causes an issue that cannot use other authentication with TLS transport encryption, and also made our confusion if we use TLS transport encryption by setting `authentication`.

## Goal

This PIP proposals add a TLS transport encryption config in client, do that can support the TLS transport encryption with any authentication.

## API Changes

- Add new methods in `org.apache.pulsar.client.api.ClientBuilder`

```java
public interface ClientBuilder extends Serializable, Cloneable {
    /**
     * Set the path to the TLS key file.
     *
     * @param tlsKeyFilePath
     * @return the client builder instance
     */
    ClientBuilder tlsKeyFilePath(String tlsKeyFilePath);

    /**
     * Set the path to the TLS certificate file.
     *
     * @param tlsCertificateFilePath
     * @return the client builder instance
     */
    ClientBuilder tlsCertificateFilePath(String tlsCertificateFilePath);

    /**
     * The file format of the key store file.
     *
     * @param tlsKeyStoreType
     * @return the client builder instance
     */
    ClientBuilder tlsKeyStoreType(String tlsKeyStoreType);
    
    /**
     * The location of the key store file.
     *
     * @param tlsTrustStorePath
     * @return the client builder instance
     */
    ClientBuilder tlsKeyStorePath(String tlsTrustStorePath);

    /**
     * The store password for the key store file.
     *
     * @param tlsKeyStorePassword
     * @return the client builder instance
     */
    ClientBuilder tlsKeyStorePassword(String tlsKeyStorePassword);
}
```

### TLS transport encryption

We can call the `tlsKeyFilePath()`, `tlsCertificateFilePath()` and `tlsTrustCertsFilePath()` to configurate the TLS transport encryption, the code so like:

```java
PulsarClient client = PulsarClient.builder()
        .serviceUrl("pulsar+ssl://my-host:6650")
        .tlsTrustCertsFilePath("/path/to/cacert.pem")
        .tlsKeyFilePath("/path/to/client-key.pem")
        .tlsCertificateFilePath("/path/to/client-cert.pem")
        .build();
```

Call the `tlsKeyFilePath()`, `tlsTrustStorePath()` to configurate the JKS TLS transport encryption, the code so like:

```java
PulsarClient client = PulsarClient.builder()
        .serviceUrl("pulsar+ssl://my-host:6650")
        .tlsKeyFilePath("/path/key.jks")
        .tlsKeyStorePassword("hello")
        .tlsTrustStorePath("/path/trust.jks")
        .tlsTrustStorePassword("hello")
        .build();
```

### TLS transport encryption with any authentication

We can call the `tlsKeyFilePath()`, `tlsCertificateFilePath()`, `tlsTrustCertsFilePath()` and `authentication()` to configurate  the TLS transport encryption with any authentication, the code so like:

```java
PulsarClient client = PulsarClient.builder()
        .serviceUrl("pulsar+ssl://my-host:6650")
        .tlsTrustCertsFilePath("/path/to/cacert.pem")
        .tlsKeyFilePath("/path/to/client-key.pem")
        .tlsCertificateFilePath("/path/to/client-cert.pem")
        .authentication(AuthenticationTls.class.getName() /* AuthenticationToken.class.getName()*/, authParams)
        .builder()
```

Call the `tlsKeyFilePath()`, `tlsTrustStorePath()` to configurate the JKS TLS transport encryption, the code so like:

```java
PulsarClient client = PulsarClient.builder()
        .serviceUrl("pulsar+ssl://my-host:6650")
        .tlsKeyFilePath("/path/key.jks")
        .tlsKeyStorePassword("hello")
        .tlsTrustStorePath("/path/trust.jks")
        .authentication(AuthenticationTls.class.getName() /* AuthenticationToken.class.getName()*/, authParams)
        .build();
```

~For `AuthenticationTls`, we need to check the authParams, when the authParams is empty, we need to read TLS config from `ClientBuilder`, otherwise read from the authParams, the authParams can override the config from `ClientBuilder`.~

When using `AuthenticationTls` and TLS config of `ClientBuilder`, we only use the `AuthenticationTls`, which can override the top level configuration. 

**Sort priority:** `AuthenticationTls` > TLS config of `ClientBuilder`.

### Plan test

- Verify TLS transport encryption without authentication
- Verify TLS transport encryption with token authentication
- Verify TLS transport encryption with TLS authentication
- Verify JKS TLS transport encryption without authentication
- Verify JKS TLS transport encryption with token authentication
- Verify JKS TLS transport encryption with TLS authentication 

### Compatibility

This PIP doesn't break TLS transport encryption and TLS authentication config, the users can still use the old config in the client.

### PR DESCRIPTION
**Pull request by nodece**
Fixes #15289 

### Motivation

The client supports TLS transport encryption and TLS authentication, this code so like:

```java
PulsarClient client = PulsarClient.builder()
                .serviceUrl("pulsar+ssl://localhost:6651")
                .tlsTrustCertsFilePath("/path/to/cacert.pem")
                .authentication(AuthenticationTls.class.getName(), authParams)
                .build()
```

We are using AuthenticationTls to perform the TLS transport encryption and TLS authentication, this causes an issue that cannot use other authentication with TLS transport encryption, and also made our confusion if we use TLS transport encryption by setting `authentication`.

### Documentation

Check the box below or label this PR directly.

Need to update docs? 

- [x] `doc-required` 
(Your PR needs to update docs and you will update later)

### CODE REVIEW

File: pulsar-common/src/main/java/org/apache/pulsar/common/util/NettyClientSslContextRefresher.java [hunk 1]
```diff
@@ -46,6 +46,8 @@ public class NettyClientSslContextRefresher extends SslContextAutoRefreshBuilder
 
     public NettyClientSslContextRefresher(SslProvider sslProvider, boolean allowInsecure,
                                           String trustCertsFilePath,
+                                          String certsFilePath,
+                                          String keyFilePath,
```
[the remark this answers is not in the document]
Comment (nodece) [reply pulsar-common/src/main/java/org/apache/pulsar/common/util/NettyClientSslContextRefresher.java:50]: This PR just improves the client SDK, we have the built-in client/admin in the broker/proxy, these also need this improvement.

File: pulsar-broker/src/test/java/org/apache/pulsar/client/impl/KeyStoreTlsProducerConsumerTestWithAuthTest.java [hunk 2]
```diff
@@ -92,6 +101,12 @@ protected void internalSetUpForBroker() {
         conf.setAuthorizationEnabled(true);
         Set<String> providers = new HashSet<>();
         providers.add(AuthenticationProviderTls.class.getName());
+
+        Properties properties = new Properties();
+        properties.setProperty("tokenSecretKey", JWT_SECRET_PATH);
```
[the remark this answers is not in the document]
Comment (nodece) [reply pulsar-broker/src/test/java/org/apache/pulsar/client/impl/KeyStoreTlsProducerConsumerTestWithAuthTest.java:106]: Great catch!

File: pulsar-common/src/main/java/org/apache/pulsar/common/util/NettyClientSslContextRefresher.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (BewareMyPower) [reply pulsar-common/src/main/java/org/apache/pulsar/common/util/NettyClientSslContextRefresher.java:50]: But it looks like the built-in client or admin doesn't pass any non-null arguments to these two arguments as well. Why do they need this two extra parameters? I mean, all references pass `null` arguments, they are meaningless.

[the remark this answers is not in the document]
Comment (nodece) [reply pulsar-common/src/main/java/org/apache/pulsar/common/util/NettyClientSslContextRefresher.java]: Fixed.

**Review by BewareMyPower — APPROVED** [review]

**Review by Technoboy- — APPROVED** [review]
LGTM

**Review by codelipenghui — APPROVED** [review]

**Pull request merged**